src/ophyd_async/core/_device.py

```diff
@@ -60,23 +75,25 @@ def set_name(self, name: str):
         name:
             New name to set
         """
-
+        self._name = name
         # Ensure self.log is recreated after a name change
-        if hasattr(self, "log"):
-            del self.log
+        self.log = LoggerAdapter(
+            getLogger("ophyd_async.devices"), {"ophyd_async_device_name": self.name}
+        )
+        for child_name, child in self.children():
+            _setup_child(self, child_name, child)
 
-        self._name = name
-        for attr_name, child in self.children():
-            child_name = f"{name}-{attr_name.rstrip('_')}" if name else ""
-            child.set_name(child_name)
-            child.parent = self
+    def __setattr__(self, name: str, value: Any) -> None:
+        if name != "parent" and isinstance(value, Device):
+            _setup_child(self, name, value)
+        return super().__setattr__(name, value)
```

Comment:
With this, a device can be an attribute in multiple parents, but it'll get it's name and `parent` field from whichever parent happens to `setattr` last? We should maybe outlaw this? (The below suggestion wouldn't work alongside the above review comment).

```suggestion
    def __setattr__(self, name: str, value: Any) -> None:
        if name == "parent" and getattr(self, "parent", None) is not None:
            raise TypeError("Can't overwrite a device's parent after it's been assigned.")
        elif isinstance(value, Device):
            _setup_child(self, name, value)
        return super().__setattr__(name, value)
```

The side effect would be that you couldn't

``` python
def __init__(self: Device, other_sub_device: Device):
    self.other_sub_device = other_sub_Device
```

where `other_sub_device` is a child in another tree.

Reply:
Alternatively if we don't want to deal with this now, we could just change the above to only `_setup_child` if it hasn't been done already, so you could still do
``` python
def __init__(self: Device, other_sub_device: Device):
    self.other_sub_device = other_sub_Device
```
But `other_sub_device` will keep its original parent and name.